### PR TITLE
Added Property relation to support conditional UI control of properties Backend feature required for #887

### PR DESCRIPTION
**Pull request by dhirajsb**

### CODE REVIEW

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@SuppressWarnings("immutables")
+public interface PropertyRelation extends Serializable {
+
+    class Builder extends ImmutablePropertyRelation.Builder {
+        // make ImmutablePropertyRelation.Builder accessible
+    }
+
+    @Value.Immutable
+    @JsonDeserialize(builder = PropertyRelationWhen.Builder.class)
+    interface PropertyRelationWhen {
```
Comment (rhuss) [inline app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: Can we avoid inner interfaces with immutables ? IDEA doesn't work with them. Alternatively, if you find a way how to compile this from within IDEA, the better ;-)

Actually, I have no idea what `PropertyRelationWhen` means. Could you elaborate a bit ? Actually, I would expect some date here, but is probably not.

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        String getId();
+
+        String getValue();
+    }
+
+    String getAction();
+
+    List<PropertyRelationWhen> getWhen();
```
Comment (rhuss) [inline app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: Should it be then `WhenList` ?

Really not sure about the naming. At least its not very intuitive.

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java (hunk 1 above)
Comment (rhuss) [reply app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: Hmm. not sure about the IDEA thing, but for `Meta` I had to extract an inner interface to make IDEA happy. But I see that there are also other inner interfaces used, so might be not an issue here.

But when keeping it, then I would remove teh `PropertyRelation` prefix, as `PropertyRelation.PropertyRelationWhen` is redundant.

File: app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java [hunk 3]
```diff
@@ -77,6 +78,22 @@
...
+    @JsonDeserialize(builder = ImmutableRelationWhen.Builder.class)
+    interface RelationWhen {
+        String getId();
+        String getValue();
+    }
+
+    @Value.Immutable
+    @Value.Style(builder = "new")
+    @JsonDeserialize(builder = ImmutableRelation.Builder.class)
+    interface Relation {
```
Comment (rhuss) [inline app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java]: What's the difference between a `Relation` and a `PropertyRelation` ?

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java (hunk 1 above)
Comment (seanforyou23) [reply app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: I could be mistaken, but I believe PropertyRelationWhen may correspond to the structure found here - https://github.com/udos86/ng-dynamic-forms/#related-form-controls, which is something we need to help out the forms interactivity in the UI. Seeing that may make sense out of some of this :)

**Comment by dhirajsb** [comment]
@rhuss or @lburgazzoli care to review this guys? Just in case I've done something obviously incorrect.

**Comment by dhirajsb** [comment]
@rhuss just saw your comments.

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java (hunk 1 above)
Comment (dhirajsb) [reply app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: What @seanforyou23 said about what the properties mean. For the name, I'm ok with renaming the inner classes to remove the redundant PropertyRelation prefix.

File: app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java (hunk 2 above)
Comment (dhirajsb) [reply app/common/model/src/main/java/io/syndesis/common/model/connection/PropertyRelation.java]: It's the way UI asked for it to be. Field names make more sense when viewing an instance data model though.

File: app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java (hunk 3 above)
Comment (dhirajsb) [reply app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java]: Good question. We basically have duplicate models for `ComponentDefinition` and `ConnectorDescriptor`. I could rename `Relation` in Component definition to `PropertyRelation` to make it similar.

File: app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java [hunk 4]
```diff
@@ -98,6 +115,7 @@
         Optional<String> getPrefix();
         Optional<String> getMultiValue();
         Optional<String> getEnumValues();
+        Optional<List<PropertyRelation>> getRelation();
```
Comment (lburgazzoli) [inline app/integration/component-proxy/src/main/java/io/syndesis/integration/component/proxy/ComponentDefinition.java]: Making a collection optional is ugly, getRelation should return an empty collection by default

**Pull request merged**